### ISSUE TITLE
Misfit Computing

### ISSUE THREAD

**Issue by MariusLTU** [clberube/BISIP#3]
Is it possible to have a function that calculates the misfit of the data once the model has be drawn?

**Comment by clberube** [comment]
Hello,
For now you can track the likelihood or model deviance, but that value is not really intuitive for misfit interpretation.
I have been using a function to compute the [Normalized Root Mean Squared Error (NRMSE)](https://en.wikipedia.org/wiki/Root-mean-square_deviation#Normalized_root-mean-square_deviation) of both the real and imaginary parts of the final average model in the development branch of bisip. It is the RMSE between observations and model, divided by the range of observations. The output is written in the CSV file with other parameters.
I will push an update to the main branch and to ```pip``` a bit later. I will close this when it is done.
c.

**Comment by MariusLTU** [comment]
Hi,
Sorry, I don't really understand how the MCMC algorithm is used here,
Correct me if I'm wrong.
When doing the inversion: for each iteration, a model curve is generated that tries to fit as much as possible the input data (by the way, is the fitting done using the least mean squares method?).
Therefore for 2000 MCMC iterations, 2000 models and parameters are provided, is it how it works?

In that case, how to be sure that the parameters estimated in the csv files are optimal?

What I would like to do in my project is to calculate the misfit for each model provided by BISIP.
The best would be to have a file with the traces of every model iterations (R0, m, t, c ...) along with the Root Mean square error for each iteration (and other estimations such as 95% confidence interval).

In that way, I could see in a table file which ones of the resulting models have the lowest RMSE.

Cheers,
Marius

**Comment by clberube** [comment]
As any statistician will tell you, with MCMC there are no way to be 100% sure that the average model found is the global optimum. However if you always converge to a similar model there are good changes that this model is a good one. Least-squares fitting doesn't make much sense for a probabilistic inversion. At each iteration the model quality is evaluated using the likelihood function. Think of it as a probabilistic misfit function.

What the MCMC algorithm does is look at the likelihood for many iterations and change the model to find a multitude of optimal results. Near the end of the simulation, if it is let run long enough, all iterations are more or less equally probable. Therefore it does not make sense to chose only 1 of them. The spread of those probable models depends only on the data noise. What makes sense is considering all iterations using basic statistics such as mean, SD, 95% HPD.

What you're asking is possible, and bisip can compute a deterministic misfit for each iteration. This information is easiest to extract when working from a python console. Using the GUI will always be much less flexible and limited to whatever free time I have to develop it.

I'll see if I can add a switch to add this information automatically in the CSV, but I don't like the idea of generating very large files automatically. 

In summary: 
-Use a simple least-squares algorithm (deterministic) if looking for only 1 solution and no uncertainty. 
-Use the MCMC algorithm in bisip (stochastic) if looking for what solution on average is the best and what the uncertainty of this solution is.

See this figure that compares the solutions obtained from both types of algorithm.
<img width="373" alt="examplefit_k389369" src="https://user-images.githubusercontent.com/20193582/37664833-5acd4b26-2c32-11e8-8904-70271200ce5b.png">

**Comment by MariusLTU** [comment]
Thanks ! I have a better idea of the MCMC now.

I'm asking you this question because I want to select all the models that are considered relevant to the data and compare them by doing PCA (Principal Component Analysis) with a geophysical software.

**Comment by clberube** [comment]
I have an idea on how to implement what you need.
This will be a feature of the new save_csv_traces function that you proposed earlier.

The function will save the modelled data (real and imaginary parts at all frequencies) and misfit values for the real and imaginary parts, in addition to the parameters used in the model for every iteration.

The results will be a CSV file like:

Iteration number  | m1 | m2 | tau1 | tau2 | c1 | c2 | R0 |  misfit.real | misfit.imag | model_freq1.real | model_freq2.real | ... | model_freq1.imag | model_freq2.imag | ... |
 --------- | ---------- | -----  |--------- | ---------- | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | ----- 
1  | --------- | ---------- | -----  | ---------- | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | ----- 
2  | --------- | ---------- | -----  |---------- | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | ----- 
...  |  --------- | ---------- | -----  |---------- | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | ----- 
N  |  --------- | ---------- | -----  |---------- | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | -----  | ----- 


That way you can reconstruct the model that was tested at every iteration. 

Stay in touch, I will push an update when I am done testing.

**Comment by MariusLTU** [comment]
That's great ! 
Thanks !

**Comment by clberube** [comment]
Salut Marius,

Update to bisip **0.0.15** with `pip` to try it. 

In the traces CSV the zmod1, zmod2, zmod3, etc. are in decreasing frequency order. Plot it to see which order makes sense. zmod1.real is of course the same frequency as zmod1.imag, and so on.

New columns for the misfits NRMSE_r (real part) and NRMSE_i (imaginary part) should be there too. Multiply the NRMSE by 100 to get in percentage.

You have to denormalize zmod and R0 parameters. To do this multiply the rho_max column by zmod and R0. The same value is used in all the rho_max column, it's the maximum measured amplitude.

Update to function `save_csv_traces` in `invResults.py`:

```python
def save_csv_traces(sol):
    """
    Saves the traces contained in mcmcinv 
    object sol to a single csv file.
    call with sol.save_csv_traces()
    """
    
    def var_depth(var):
        # Pass stochastic or deterministic
        return int(var.trace().size/len(var.trace()))
    
    # Get some basic information from the mcmc object
    MDL, filepath = sol.MDL, sol.filename
    model = get_model_type(sol)
    sample_name = filepath.replace("\\", "/").split("/")[-1].split(".")[0]

    # Decide where to save the csv
    save_where = '/TraceResults/'
    working_path = getcwd().replace("\\", "/")+"/"
    save_path = working_path + save_where + "%s/"%sample_name # Add subfolder for the sample
    
    # Get stochastics and deterministics
    sto = MDL.stochastics
    det = MDL.deterministics
    # Get names of parameters for save file
    pm_names = [s.__name__ for s in sto] + [d.__name__ for d in det]
    
    # Get all stochastic and deterministic variables
    trl = [s for s in sto] + [d for d in det]
    
    # Concatenate all traces in 1 matrix
    trace_mat = np.hstack([t.trace().reshape(-1, var_depth(t)) for t in trl])
    # Insert normalization resistivity in first column
    trace_mat = np.insert(trace_mat, 0, sol.data['Z_max'], axis=1) 

    # Get numbers for each subheader
    num_names = [var_depth(s) for s in sto] + [var_depth(d) for d in det]    
    # Make list of headers
    headers = [['%s%d'%(pm_names[p],x+1) for x in range(num_names[p])] if num_names[p] > 1 else [pm_names[p]] for p in range(len(pm_names))]
    
    # Change zmod numbers for real and imaginary parts
    for h, head in enumerate(headers):
        if 'zmod' in head[0]:   
            headers[h] = [head[i]+'.real' for i in range(int(len(head)/2))] + [head[i]+'.imag' for i in range(int(len(head)/2))]    
    
    flat_headers = [item for sublist in headers for item in sublist] # Flaten list 
    flat_headers.insert(0, 'rho_max') # Insert normalization resistivity in first column
    header = ','.join(flat_headers) # Join list into csv string 
    
    # Do the saving
    print("\nSaving CSV traces in:\n", save_path)
    if not path.exists(save_path):
        makedirs(save_path)
    np.savetxt(save_path+'TRACES_%s-%s_%s.csv' %(sol.model,model,sample_name), trace_mat, delimiter=',', header=header, comments="")
```

I also fixed the 'Preview data' button
Cheers! 
c.

**Comment by MariusLTU** [comment]
Hi Charles,
I can't open the GUI because of this error : 
"
  File "bisip\GUI.py", line 729, in make_options
    self.save_options[k].set(self.root_ini[k])

KeyError: 'Save traces figure' "

Cheers !

**Comment by MariusLTU** [comment]
Hey Charles,
Everything seems to work ! Except that now it seems that the data doesn't really fit as it used to be.

I still have one question:
On the csv files, the zmod.real and the zmod.imag correspond to the fitting of the complex resistivity as function of the frequency.
So each zmodi.real and zmodi.imag (for i = 1 to N) corresponds to the absciss of the spectra which is a frequency. So I guess zmod1 is the impedance at low frequency and zmod20 is the impedance at high frequency?

Is it possible to know the exact frequency associated with each zmod?

Cheers !
Marius

**Comment by clberube** [comment]
I think zmod1 is impedance at high frequency and zmodN at low frequency.  The exact frequencies are the same as the ones in your input data.

I didn't modify the models or fitting procedure, just added this new output to the csv file. Are you having problems with the example data?

Quite a bit of experimentation is needed to find good MCMC parameters for your own data. At our lab we have quite a large collection of complex resistivity spectra from various rock types from ore deposits and get over 90 % fit success rate with the following parameters (I use mostly 4th order polynomial Debye decomposition, and not all data can be fit with a ColeCole model):

Adaptive: True
Iterations: 100 000
Burn-in: 80 000
Thin factor: 1
Cov delay: 10 000
Cov update interval: 5 000

You can also try this if you can't fit, but that will take longer:
Adaptive: True
Iterations: 500 000
Burn-in: 400 000
Thin factor: 10
Cov delay: 50 000
Cov update interval: 50 000

or with ColeCole models the non-adaptive should still work well:
Adaptive: False
Iterations: 100 000
Burn-in: 80 000
Thin factor: 1
Tuning interval: 10 000
proposal scale: 1

Dias model isn't very flexible and will not fit perfectly, but it will converge super fast with
Adaptative: False
Iterations: 10 000
Burn-in: 8 000
Thin factor: 1
Tuning interval: 1 000
proposal scale: 1

If you can't fit after 500 000 iterations then it's not worth going longer, try to experiment with all parameters.

**Comment by MariusLTU** [comment]
Thanks a lot Charles !

Marius
